c2rust-transpile/src/c_ast/mod.rs

```diff
@@ -973,8 +973,10 @@ impl TypedAstContext {
         }
     }
 
-    pub fn sort_top_decls(&mut self) {
-        // Group and sort declarations by file and by position
+    /// Sort the top-level declarations by file and source location
+    /// so that we preserve the ordering of all declarations in each file.
+    /// This preserves the order when we emit the converted declarations.
+    pub fn sort_top_decls_for_emitting(&mut self) {
         let mut decls_top = mem::take(&mut self.c_decls_top);
         decls_top.sort_unstable_by(|a, b| {
```

Comment:
`compare_src_locs` seems to be doing a lot more than expected, but it looks fine. I'm not sure where the reordering or reversal are coming from.

Reply:
I'm moving the re-ordering to the where it's necessary (emitting) rather than converting.  This can be bundled with #1372, but clearly there's a bug in just this, so it helps to isolate the two.